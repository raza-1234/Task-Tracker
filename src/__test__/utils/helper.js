export const buildTaskPayload = (count) => {
  const tasks = [];

  for (let i=1; i <= count; i++){
    tasks.push({
      checked: false,
      date: `2024-02-0${i}`,
      description: `Fake Description ${i}`,
      id: i,
      task: `Fake Title ${i}`,
      time: `05:0${i}`
    })
  }
  return tasks;
}