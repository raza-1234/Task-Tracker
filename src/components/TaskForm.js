import React from 'react'
import { useState , useEffect} from 'react'
import api from '../axios/api';
import { useNavigate } from 'react-router-dom';

const TaskForm = ({task, taskId, setTask, setTaskId, id, fetchData}) => {
    //state for managing inputs
    const [newTask , setNewTask] = useState("");
    const [date , setDate] = useState("");
    const [time , setTime] = useState("");
    const [description , setDescription] = useState("");
    const [reminder , setReminder] = useState(false)
    const [currentTask , setCurrentTask] = useState();

    const navigate = useNavigate(); //navigate

    function getData(){
        if(task.length === 0){
            fetchData();
        } else {
            setCurrentTask(task.find(item => item.id.toString() === id))
        }
    }

    useEffect(() => {
        if(id) {
            getData();
            if(currentTask) {
                setNewTask(currentTask.task);
                setDate(currentTask.date);
                setTime(currentTask.time);
                setDescription(currentTask.description)
                setReminder(currentTask.checked);
            }
        }
    },[currentTask, task])

    async function submitHandler(event){
        event.preventDefault();

        if(newTask.trim() === "" || date.trim() === ""  || time.trim() === "" || description.trim() === "") {
          return false
        }
        try {
            const newTaskList = {id: id, task: newTask , checked : reminder , date: date , time : time , description : description };
            
            if(!id) {
                id =  task.length ? Number(task[(task.length) - 1].id) + 1 : 1;
                if(!newTaskList.id) {
                    newTaskList.id = id.toString();
                }
                const response = await api.post("/tasks" , newTaskList)
                setTask([...task , response.data])
            } else {
                const response = await api.put(`/tasks/${id}` , {...newTaskList, id: id.toString()})
                setTask(task.map((task) => {
                    return (task.id === id.toString() ? response.data : task)
                }))
                navigate("/")
            }
        } catch(err) {
            console.log(err);
        }
        setNewTask("");
        setDate("")
        setTime("")
        setDescription("")
        setReminder(false);
        setTaskId(null)
    }

  return (
    <form data-testid="taskForm" className='taskForm'  onSubmit={submitHandler}>
        <label htmlFor='task'>Task</label>
        <input
            data-testid="textInput"
            required
            type='text'
            id='task'
            placeholder='Add Task'
            value={newTask}
            onChange={(e) => setNewTask(e.target.value)}
        />

        <label htmlFor='date'>Date</label>
        <input
            data-testid="dateInput"
            required
            type='date'
            id='date'
            placeholder='Set Date'
            value={date}
            onChange = {(e) => setDate(e.target.value)}
        />

        <label htmlFor='time'>Time</label>
        <input
            data-testid="timeInput"
            required
            type='time'
            id='time'
            placeholder='Set Time'
            value={time}
            onChange = {(e) => setTime(e.target.value)}
        />

        <label htmlFor='description'>Task Description</label>
        <textarea
            type="text"
            data-testid="descriptionInput"
            required
            id='description'
            placeholder='Task Description ...'
            value={description}
            onChange = {(e) => setDescription(e.target.value)}
            rows="6"
            cols="60"/>

        <label htmlFor='Reminder Check'>Set Reminder</label>
        <input
            data-testid="reminderInput"
            className='checkBoxInput'
            id='Reminder Check'
            type='checkBox'
            checked= {reminder}
            onChange={(e) => setReminder(!reminder)}
        />

        <button className='saveTaskButton'>{id ? 'Update Task': 'Save Task'}</button>
      </form>
  )
}

export default TaskForm
