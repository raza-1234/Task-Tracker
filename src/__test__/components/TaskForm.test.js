import { render, screen, fireEvent, act, waitFor } from '@testing-library/react';
import TaskForm from "../../components/TaskForm";
import { BrowserRouter } from 'react-router-dom';
import api from "../../axios/api" 
import { buildTaskPayload } from "../utils/helper"

jest.mock("../../axios/api", () => {
  return ({
    ...jest.requireActual(),
    post: jest.fn(),
    put: jest.fn()
  })
})

describe("TaskForm", () => {
  const mockClickHandler = jest.fn();
  const mockTask = buildTaskPayload(2) 

  const mockPostData = {
    checked: false,
    date: "2024-02-01",
    description: "some description",
    id: "3", 
    task: "sample 1",
    time: "05:47"
  }

  it("should render successfully", () => {
    const { container } = render(
      <BrowserRouter>
        <TaskForm 
          task = {mockTask} 
          taskId={1}
          setTask={mockClickHandler} 
          setTaskId={mockClickHandler} 
          id={1} 
          fetchData={mockClickHandler}
        />
      </BrowserRouter>
    )

    const taskForm = screen.getByTestId("taskForm")
    expect(taskForm).toBeInTheDocument()

    expect(screen.getByLabelText("Task")).toBeInTheDocument()
    expect(screen.getByLabelText("Date")).toBeInTheDocument()
    expect(screen.getByLabelText("Time")).toBeInTheDocument()
    expect(screen.getByLabelText("Task Description")).toBeInTheDocument()
    expect(screen.getByLabelText("Set Reminder")).toBeInTheDocument()
    expect(screen.getByRole("button")).toBeInTheDocument()
  })

  it("should render input fields", () => {
    render(
      <BrowserRouter>
        <TaskForm 
          task = {mockTask} 
          taskId={1}
          setTask={mockClickHandler} 
          setTaskId={mockClickHandler} 
          id={1} 
          fetchData={mockClickHandler}
        />
      </BrowserRouter>
    )

    expect(screen.getByTestId("textInput")).toHaveAttribute("type", "text")
    expect(screen.getByTestId("dateInput")).toHaveAttribute("type", "date")
    expect(screen.getByTestId("timeInput")).toHaveAttribute("type", "time")
    expect(screen.getByTestId("descriptionInput")).toHaveAttribute("type", "text")
    expect(screen.getByTestId("reminderInput")).toHaveAttribute("type", "checkBox")
  })

  it("should render 'update task' button if there is an id", async () => {
    render(
      <BrowserRouter>
        <TaskForm 
          task = {mockTask} 
          taskId={1}
          setTask={mockClickHandler} 
          setTaskId={mockClickHandler} 
          id={1} 
          fetchData={mockClickHandler}
        />
      </BrowserRouter>
    )

    const updateButton = screen.getByRole("button", {name:"Update Task"})
    expect(updateButton).toBeInTheDocument()
  })

  it("should render 'save task' button if no id exists.", async () => {
    render(
      <BrowserRouter>
        <TaskForm 
          task = {mockTask} 
          taskId={1}
          setTask={mockClickHandler} 
          setTaskId={mockClickHandler} 
          fetchData={mockClickHandler}
        />
      </BrowserRouter>
    )

    const updateButton = screen.getByRole("button", {name:"Save Task"})
    expect(updateButton).toBeInTheDocument()
  })

  it("should post data and return that data", async () => {
    api.post.mockResolvedValue({data: mockPostData})

    let container;
    await act( async () => container = render(
      <BrowserRouter>
        <TaskForm 
          task = {mockTask} 
          taskId={1}
          setTask={mockClickHandler} 
          setTaskId={mockClickHandler} 
          fetchData={mockClickHandler}
        />
      </BrowserRouter>
    ));

    const button = screen.getByRole("button", {name: "Save Task"})

    const task = screen.getByLabelText("Task");
    const date = screen.getByLabelText("Date")
    const time = screen.getByLabelText("Time")
    const description = screen.getByLabelText("Task Description")
    const reminder = screen.getByLabelText("Set Reminder")

      fireEvent.change(task, {target: {value: "sample 1"}})
      fireEvent.change(date, { target: { value: '2024-02-01' } });
      fireEvent.change(time, {target: {value: "05:47"}})
      fireEvent.change(description, {target: {value: "some description"}})
      fireEvent.change(reminder, {target: {value: false}})

    expect(screen.getByDisplayValue("sample 1")).toBeInTheDocument()
    fireEvent.click(button); 
    
    await waitFor(() => {
      expect(container.container).toMatchSnapshot();
      expect(api.post).toHaveBeenCalledWith("/tasks", mockPostData) 
    }) 
  });

  it("should update data and return that data", async () => {
    api.put.mockResolvedValue({data: mockPostData})

    let container;
    await act( async () => container = render(
      <BrowserRouter>
        <TaskForm 
          task = {mockTask} 
          taskId={1}
          setTask={mockClickHandler} 
          id={3}
          setTaskId={mockClickHandler} 
          fetchData={mockClickHandler}
        />
      </BrowserRouter>
    ));

    const button = screen.getByRole("button", {name: "Update Task"})

    const task = screen.getByLabelText("Task");
    const date = screen.getByLabelText("Date")
    const time = screen.getByLabelText("Time")
    const description = screen.getByLabelText("Task Description")
    const reminder = screen.getByLabelText("Set Reminder")

    fireEvent.change(task, {target: {value: "sample 1"}})
    fireEvent.change(date, { target: { value: '2024-02-01' } });
    fireEvent.change(time, {target: {value: "05:47"}})
    fireEvent.change(description, {target: {value: "some description"}}) 
    fireEvent.change(reminder, {target: {value: false}})
    fireEvent.click(button); 
    
    await waitFor(() => {
      expect(container.container).toMatchSnapshot();
      expect(api.put).toHaveBeenCalledWith("/tasks/3", mockPostData)  
    }) 
  }); 
 
}) 