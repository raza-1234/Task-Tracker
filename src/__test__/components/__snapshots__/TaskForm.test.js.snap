// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm should post data and return that data 1`] = `
<div>
  <form
    class="taskForm"
    data-testid="taskForm"
  >
    <label
      for="task"
    >
      Task
    </label>
    <input
      data-testid="textInput"
      id="task"
      placeholder="Add Task"
      required=""
      type="text"
      value="sample 1"
    />
    <label
      for="date"
    >
      Date
    </label>
    <input
      data-testid="dateInput"
      id="date"
      placeholder="Set Date"
      required=""
      type="date"
      value="2024-02-01"
    />
    <label
      for="time"
    >
      Time
    </label>
    <input
      data-testid="timeInput"
      id="time"
      placeholder="Set Time"
      required=""
      type="time"
      value="05:47"
    />
    <label
      for="description"
    >
      Task Description
    </label>
    <textarea
      cols="60"
      data-testid="descriptionInput"
      id="description"
      placeholder="Task Description ..."
      required=""
      rows="6"
      type="text"
    >
      some description
    </textarea>
    <label
      for="Reminder Check"
    >
      Set Reminder
    </label>
    <input
      class="checkBoxInput"
      data-testid="reminderInput"
      id="Reminder Check"
      type="checkBox"
      value="false"
    />
    <button
      class="saveTaskButton"
    >
      Save Task
    </button>
  </form>
</div>
`;

exports[`TaskForm should update data and return that data 1`] = `
<div>
  <form
    class="taskForm"
    data-testid="taskForm"
  >
    <label
      for="task"
    >
      Task
    </label>
    <input
      data-testid="textInput"
      id="task"
      placeholder="Add Task"
      required=""
      type="text"
      value="sample 1"
    />
    <label
      for="date"
    >
      Date
    </label>
    <input
      data-testid="dateInput"
      id="date"
      placeholder="Set Date"
      required=""
      type="date"
      value="2024-02-01"
    />
    <label
      for="time"
    >
      Time
    </label>
    <input
      data-testid="timeInput"
      id="time"
      placeholder="Set Time"
      required=""
      type="time"
      value="05:47"
    />
    <label
      for="description"
    >
      Task Description
    </label>
    <textarea
      cols="60"
      data-testid="descriptionInput"
      id="description"
      placeholder="Task Description ..."
      required=""
      rows="6"
      type="text"
    >
      some description
    </textarea>
    <label
      for="Reminder Check"
    >
      Set Reminder
    </label>
    <input
      class="checkBoxInput"
      data-testid="reminderInput"
      id="Reminder Check"
      type="checkBox"
      value="false"
    />
    <button
      class="saveTaskButton"
    >
      Update Task
    </button>
  </form>
</div>
`;
